feat(formatting): Add memory units adaptive formatter to format bytes

mkopec87 opened the pull request:
Add new memory units adaptive formatter to format bytes units into kB, MB, GB, etc.

### SUMMARY
Memory unit (bytes) is typically formatted into kB (kilobytes), MB (megabytes), etc.
Superset formatting options don't enable adaptive formatting with proper unit naming in case of bytes values.
There are two most commonly used multi-byte unit types (according to [wikipedia](https://en.wikipedia.org/wiki/Byte)), this PR implements them both.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
![image](https://github.com/user-attachments/assets/e619b401-1df4-46a0-8fb8-86305ba21be8)

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->
1. Create a Chart, for example Table, with at least one numeric data column.
2. Edit the Chart, in the "Customize" tab select one of "Memory in bytes" options for formatting of your numeric column.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

mkopec87 committed:
- Add new memory units adaptive formatter to format bytes units into kB, MB, GB, etc.

mkopec87 committed:
- Fix boolean type declaration
- Correct linter warnings

mkopec87 committed:
- Correct tsc error

mkopec87 committed:
Merge branch 'master' into feature/formatting-memory-units

villebro reviewed (commented):
This is a great feature, and one that many have been looking forward to seeing. In general this looks good, but would you be able to add some unit tests?

mkopec87 commented:
@villebro , sure thing, let me give it a try :) I never wrote any Javascript before, therefore happy to get some feedback :)
CI setup is very useful here.

mkopec87 committed:
Merge branch 'master' into feature/formatting-memory-units

mkopec87 committed:
- Add unit tests to MemoryFormatter
- Make decimals option a parameter of MemoryFormatter
- Handle negative values in MemoryFormatter
- Handle values exceeding maximal unit value in MemoryFormatter

mkopec87 commented:
@villebro I added unit tests and fixed some bugs on the way :) Let me know what you think!

villebro commented:
It seems there's a linting issue with the test:
<img width="693" alt="image" src="https://github.com/user-attachments/assets/2761cb59-15d2-4074-b566-fe84bab0cc93">

villebro commented on superset-frontend/packages/superset-ui-core/test/number-format/factories/createMemoryFormatter.test.ts:
```diff
@@ -0,0 +1,91 @@
+    expect(formatter(1000)).toBe('1kB');
+    expect(formatter(1111)).toBe('1.11kB');
+    expect(formatter(1024)).toBe('1.02kB');
+    expect(formatter(1337)).toBe('1.34kB');
```
lol

villebro commented on superset-frontend/packages/superset-ui-core/src/number-format/factories/createMemoryFormatter.ts:
```diff
@@ -0,0 +1,55 @@
+      if (value === 0) return '0B';
+
+      const sign = value > 0 ? '' : '-';
+      value = Math.abs(value);
```
Just create a new `const` and use that to get around the linting error:
```suggestion
      const absValue = Math.abs(value);
```

mkopec87 replied:
changed as suggested

villebro reviewed (commented):
LGTM after the linting issue is fixed

villebro commented on superset-frontend/packages/superset-ui-core/test/number-format/factories/createMemoryFormatter.test.ts:
```diff
@@ -0,0 +1,91 @@
+
+import { NumberFormatter, createMemoryFormatter } from '@superset-ui/core';
+
+describe('createMemoryFormatter()', () => {
+  it('creates an instance of MemoryFormatter', () => {
+    const formatter = createMemoryFormatter();
+    expect(formatter).toBeInstanceOf(NumberFormatter);
+  });
+  it('formats bytes in human readable format with default options', () => {
```
Oh, one more thing, we try to steer away from nested tests. So instead of doing `describe('foo)` with many `it('bar')` and `it('baz')`, just do individual `test('foo bar')` and `test('foo baz')`

mkopec87 replied:
sure, done

mkopec87 committed:
Merge branch 'apache:master' into feature/formatting-memory-units

mkopec87 committed:
- Unnest tests of MemoryFormatter
- Avoid function parameter reassignment in MemoryFormatter

villebro reviewed (approved):
Looks great, fingers crossed for CI to pass 🤞 Only comment is it'd be great to be able to customize the decimals, but that's outside the scope of this PR. I believe there's a discussion about this lying around somewhere, and would be a great improvement (the idea would be to be able to provide arbitrary config objects to the formatters), but that would require a SIP, so let's leave that for a future improvement.

mkopec87 commented:
Thanks @villebro for your help merging this :)

WilliamGeuns1 commented:
@villebro any idea when this will be in the release?

rusackas commented:
As far as I know it didn't make the cut opportunity for 4.1, so it'd most likely be in either a 4.2 (if that happens) or 5.0.
